### PR TITLE
VLM Data Collator - Make text & image mixing work efficiently

### PR DESCRIPTION
**Pull request by mmathew23**
Currently finetuning VLM with mixed text and text/image rows will not work. The processing class complains if you pass non images, and even if you pass that the vision models requires full tenors for every batch. 

This PR adds a custom data sampler to solve for this. Some alternatives were forcing single batch training, sub batching the mixed batches, and passing dummy images.

Dummy images would force extra flops and masking is requires to make sure the dummy image doesn't influence the model. Not efficient and hard to maintain for new models with new masking strategies.

Single batch woudl work, but that feels to strict. And sub batching would require patching the VLM to loop over image vs non image sub batches. Could be viable but still not as efficient, and harder to extend to multi gpu cases.

Instead I opted for a data sampler to control the order of batches. It will yield batches with all images+text or only text. The model will entirely bypass the vision model so it's most efficient. [This colab notebook](https://colab.research.google.com/drive/1vPE8KY84E2SHaX_LM2DrmJahZB4e5pyj?usp=sharing) was tested for llama, qwen, and pixtral. 

We also note that setting truncation to True was causing training errors (reproducible with the unsloth vlm notebook) and that is no longer set. In order to pass a sampler I follow how Trainer's inner training loop get's patched, and patch training sampler to check for a now config argument, group_mixed_image_text.

### CODE REVIEW

**Comment by mmathew23** [comment]
Also depends on unsloth [PR #2133](https://github.com/unslothai/unsloth/pull/2133)

**Comment by rolandtannous** [comment]
People can still create their custom collator. I would moreover add that when you have a special case like this, you preprocess in the Dataset Class (People are not even creating dataset classes..)
But if people come to a consensus that going ahead with a collator to support mixed datasets is the right way, I suggest that a new MixedDataCollator class is created instead of trying to patch the VisionDataCollator itself to work with mixed datasets. That would be a slightly more elegant approach and a better separation of concerns approach.

**Comment by mmathew23** [comment]
That's reasonable. 

I was working under the assumption that we should be fixing the current collator to support mixed datasets. But yea it would be good to get an opinion on which way to go.
